app/src/main/java/com/duckduckgo/app/bookmarks/migration/AppDatabaseBookmarksMigrationCallback.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.bookmarks.migration
+
+import androidx.room.RoomDatabase
+import androidx.sqlite.db.SupportSQLiteDatabase
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.db.AppDatabase
+import com.duckduckgo.sync.store.Entity
+import com.duckduckgo.sync.store.EntityType.BOOKMARK
+import com.duckduckgo.sync.store.EntityType.FOLDER
+import com.duckduckgo.sync.store.Relation
+import dagger.Lazy
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.asExecutor
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class AppDatabaseBookmarksMigrationCallback(
+    private val appDatabase: Lazy<AppDatabase>,
+    private val dispatcherProvider: DispatcherProvider,
+) : RoomDatabase.Callback() {
+
+    override fun onOpen(db: SupportSQLiteDatabase) {
```

Comment:
better than version, we can check if the folder is empty or not and act accordingly. at some point we will need to delete old bookmarks code, and this will be another of the things to remove.